### PR TITLE
Add Webhook for Repository ruleset

### LINKED ISSUE #3295: Feature: repository ruleset event type not defined
From https://github.com/cbrgm/githubevents/issues/94#issuecomment-2378597876 - GitHub are now creating repository rulesets by default rather than branch protection rules (via the UI at least).

[API definition](https://docs.github.com/en/webhooks/webhook-events-and-payloads#repository_ruleset).

### PR DESCRIPTION
**Pull request by unamdev0**
Added  Repository ruleset event 
Fixes #3295

### CODE REVIEW

**Comment by unamdev0** [comment]
@gmlewis  please ignore the previous PR and review this. This PR contains just my code and hasn't been reabased with master branch.
 updated `RuleSet` to `Ruleset` everywhere, also improved the comment describing the repository ruleset event

**Review by gmlewis — CHANGES_REQUESTED** [review]
I'm going to hold off reviewing this one until comments from the last PR are addressed.
Specifically,

> Godocs should always use complete sentences with punctuation so that the auto-generated package documentation formats and reads nicely.

It looks like GitHub ate my other lengthy comment, unfortunately... but the gist of it was that every exported struct needs a Godoc-style comment to explain its purpose and usage for the auto-generated package documentation.

**Comment by unamdev0** [comment]
@gmlewis  please go ahead and review the code, I've added comments for all the new structs generated

File: github/repos_rules.go [hunk 1]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+}
+
+// RepositoryRulesetEditedSource represents a source change for the ruleset.
+type RepositoryRulesetEditedSource struct {
+	From *string `json:"from,omitempty"`
+}
+
+// RepositoryRulesetEditedSources represents multiple source changes for the ruleset.
+type RepositoryRulesetEditedSources struct {
+	From []*string `json:"from,omitempty"`
```
Comment (gmlewis) [inline github/repos_rules.go:186]:
```suggestion
	From []string `json:"from,omitempty"`
```

File: github/repos_rules.go [hunk 2]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+	SourceType *string `json:"source_type,omitempty"`
+	Source     string  `json:"source"`
+	// Possible values for enforcement: "disabled", "active", "evaluate"
+	Enforcement  string         `json:"enforcement"`
+	BypassActors []*BypassActor `json:"bypass_actors,omitempty"`
+	// Possible values for current user can bypass: "always", "pull_requests_only", "never"
+	CurrentUserCanBypass *string                  `json:"current_user_can_bypass,omitempty"`
+	NodeID               *string                  `json:"node_id,omitempty"`
+	Links                *RepositoryRulesetLink   `json:"_links,omitempty"`
+	Conditions           *json.RawMessage         `json:"conditions,omitempty"`
```
Comment (gmlewis) [inline github/repos_rules.go:251]:
```suggestion
	Conditions           json.RawMessage         `json:"conditions,omitempty"`
```

File: github/repos_rules.go [hunk 3]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+// RepositoryRulesetCodeScanningRule defines a code scanning rule for the repository.
+type RepositoryRulesetCodeScanningRule struct {
+	//Value for Type:"code_scanning"
+	Type       string                      `json:"type"`
+	Parameters *RuleCodeScanningParameters `json:"parameters,omitempty"`
+}
+
+// RuleCodeScanningParameters defines parameters for code scanning rules.
+type RuleCodeScanningParameters struct {
+	CodeScanningTools []CodeScanningTool `json:"code_scanning_tools"`
```
Comment (gmlewis) [inline github/repos_rules.go:381]:
```suggestion
	CodeScanningTools []*CodeScanningTool `json:"code_scanning_tools,omitempty"`
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @unamdev0 !
Just a few linter errors to clean up, please, and a few tweaks, then we should be ready for a second LGTM+Approval from any other contributor to this repo before merging.

**Comment by unamdev0** [comment]
Done @gmlewis, all the suggestion made have been added and test files have been modified accordingly

File: github/repos_rules.go [hunk 4]
```diff
@@ -378,7 +378,7 @@ type RepositoryRulesetCodeScanningRule struct {
 
 // RuleCodeScanningParameters defines parameters for code scanning rules.
 type RuleCodeScanningParameters struct {
-	CodeScanningTools []CodeScanningTool `json:"code_scanning_tools"`
+	CodeScanningTools []CodeScanningTool `json:"code_scanning_tools,omitempty"`
```
Comment (gmlewis) [inline github/repos_rules.go:381]:
```suggestion
	CodeScanningTools []*CodeScanningTool `json:"code_scanning_tools,omitempty"`
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
It looks like there are a few more linter errors - you can see them in the GitHub user interface, I believe.

File: github/repos_rules.go (hunk 4 above)
Comment (unamdev0) [reply github/repos_rules.go:381]: sorry missed this, correcting it right away

**Comment by unamdev0** [comment]
Hi @gmlewis, I'm not able to see any linting issues shown in github interface, when running script/lint.sh. Other than that,I've made the change that was missed last time


Screenshot 2024-10-14 at 8 50 40 PM

**Comment by gmlewis** [comment]
Ah, to see them locally you probably need to merge the latest `master` into your PR... but I'll rerun the workflows and see if I can take screenshots for you.

**Comment by gmlewis** [comment]
@unamdev0 - can you see this?
https://github.com/google/go-github/actions/runs/11330493244/job/31508792189?pr=3305

File: github/repos_rules.go [hunk 5]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
 	RulesetID         int64            `json:"ruleset_id"`
 }
 
+// RepositoryRulesetEditedChanges represents  the changes made to a repository ruleset
```
Comment (gmlewis) [inline github/repos_rules.go:171]:
```suggestion
// RepositoryRulesetEditedChanges represents  the changes made to a repository ruleset.
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
I've gone ahead and made suggestions that should clear up the linter errors.

**Comment by unamdev0** [comment]
@gmlewis  added the suggestions, should I add the changes in generated files manually as well as shown in the github UI?

**Comment by gmlewis** [comment]
> @gmlewis added the suggestions, should I add the changes in generated files manually as well as shown in the github UI?

No, you should never edit the auto-generated files.  If you find problems in them (which I don't think there are any), then the code that generates them `gen-*.go` should be modified... but usually, users of this repo can leave maintenance of those up to the maintainers of the repo.  Is that what you were asking?

**Comment by gmlewis** [comment]
Can you please push those changes to the suggested changes?  I'm not seeing them yet.

**Comment by unamdev0** [comment]
> No, you should never edit the auto-generated files. If you find problems in them (which I don't think there are any), then the code that generates them `gen-*.go` should be modified... but usually, users of this repo can leave maintenance of those up to the maintainers of the repo. Is that what you were asking?

Oh okay, I've pushed changes with period at the end of comments, and not modified any other file, can you please review once

File: github/repos_rules.go [hunk 6]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+}
+
+// RepositoryRulesetRuleType represents the type of a ruleset rule.
+type RepositoryRulesetRuleType struct {
+	Type string `json:"type"`
+}
+
+// RepositoryRulesetUpdateRule defines an update rule for the repository.
+type RepositoryRulesetUpdateRule struct {
+	//Value for Type: "update"
```
Comment (alexandear) [inline github/repos_rules.go:295]: I propose to change:
```suggestion
	// Type can be one of: "update".
```

This is similar to what we already have for the `Issue.State` comment: https://github.com/google/go-github/blob/8ea16b939809ee06e01354f8923c0ace033217fb/github/issues.go#L31

Comment (alexandear) [reply github/repos_rules.go:295]: The comment applies to all added comments.

File: github/event_types_test.go [hunk 7]
```diff
@@ -9506,6 +9506,1782 @@ func TestReleaseEvent_Marshal(t *testing.T) {
 	testJSONMarshal(t, u, want)
 }
 
+func TestRepositoryRulesetEvent_Marshal(t *testing.T) {
```
Comment (gmlewis) [inline github/event_types_test.go:9509]:
```suggestion
func TestRepositoryRulesetEvent_Marshal(t *testing.T) {
	t.Parallel()
```

**Review by gmlewis — CHANGES_REQUESTED** [review]

**Comment by gmlewis** [comment]
@unamdev0 - I see what the problem is here.  
We made some updates to this repo since the time that you started creating it.
You need to fix it on your side by getting the latest generators.
I've already merged in the latest from the `master` branch, so the following should be all you need to do to update the PR:

```bash
git checkout feature/add-repositoryrulesetevent
git pull origin feature/add-repositoryrulesetevent
go generate ./...
git commit -am 'Run latest go generate'
git push origin feature/add-repositoryrulesetevent
```

(Instead of `go generate ./...` you can run any/all of the scripts listed in step 4 of CONTRIBUTING.md including `./script/generate.sh`)

File: github/repos_rules.go [hunk 8]
```diff
@@ -292,36 +292,36 @@ type RepositoryRulesetRuleType struct {
 
 // RepositoryRulesetUpdateRule defines an update rule for the repository.
 type RepositoryRulesetUpdateRule struct {
-	//Value for Type: "update"
+	//Type can be one of: "update".
```
Comment (gmlewis) [inline github/repos_rules.go:295]:
```suggestion
	// Type can be one of: "update".
```

File: github/repos_rules.go [hunk 9]
```diff
@@ -292,36 +292,36 @@ type RepositoryRulesetRuleType struct {
...
 	Type       string                                   `json:"type"`
 	Parameters *UpdateAllowsFetchAndMergeRuleParameters `json:"parameters,omitempty"`
 }
 
 // RepositoryRulesetMergeQueueRule defines a merge queue rule for the repository.
 type RepositoryRulesetMergeQueueRule struct {
-	//Value for Type: "merge_queue"
+	//Type can be one of: "merge_queue".
```
Comment (gmlewis) [inline github/repos_rules.go:302]:
```suggestion
	// Type can be one of: "merge_queue".
```

File: github/repos_rules.go [hunk 10]
```diff
@@ -292,36 +292,36 @@ type RepositoryRulesetRuleType struct {
...
 	Type       string                    `json:"type"`
 	Parameters *MergeQueueRuleParameters `json:"parameters,omitempty"`
 }
 
 // RepositoryRulesetRequireDeploymentsRule defines a rule for required deployments.
 type RepositoryRulesetRequireDeploymentsRule struct {
-	//Value for Type: "required_deployments"
+	//Type can be one of: "required_deployments".
```
Comment (gmlewis) [inline github/repos_rules.go:309]:
```suggestion
	// Type can be one of: "required_deployments".
```

File: github/repos_rules.go [hunk 11]
```diff
@@ -292,36 +292,36 @@ type RepositoryRulesetRuleType struct {
...
 	Type       string                                        `json:"type"`
 	Parameters *RequiredDeploymentEnvironmentsRuleParameters `json:"parameters,omitempty"`
 }
 
 // RepositoryRulesetPullRequestRule defines a rule for pull requests.
 type RepositoryRulesetPullRequestRule struct {
-	//Value for Type: "pull_request"
+	//Type can be one of: "pull_request".
```
Comment (gmlewis) [inline github/repos_rules.go:316]:
```suggestion
	// Type can be one of: "pull_request".
```

File: github/repos_rules.go [hunk 12]
```diff
@@ -292,36 +292,36 @@ type RepositoryRulesetRuleType struct {
...
 
 	Type       string                     `json:"type"`
 	Parameters *PullRequestRuleParameters `json:"parameters,omitempty"`
 }
 
 // RepositoryRulesetRequiredStatusChecksRule defines a rule for required status checks.
 type RepositoryRulesetRequiredStatusChecksRule struct {
-	//Value for Type: "required_status_checks"
+	//Type can be one of: "required_status_checks".
```
Comment (gmlewis) [inline github/repos_rules.go:324]:
```suggestion
	// Type can be one of: "required_status_checks".
```

File: github/repos_rules.go [hunk 13]
```diff
@@ -335,43 +335,43 @@ type RepositoryRulesetPatternRule struct {
 
 // RepositoryRulesetFilePathRestrictionRule defines a file path restriction rule for the repository.
 type RepositoryRulesetFilePathRestrictionRule struct {
-	//Value for Type: "file_path_restriction"
+	//Type can be one of: "file_path_restriction".
```
Comment (gmlewis) [inline github/repos_rules.go:338]:
```suggestion
	// Type can be one of: "file_path_restriction".
```

File: github/repos_rules.go [hunk 14]
```diff
@@ -335,43 +335,43 @@ type RepositoryRulesetPatternRule struct {
...
 	Type       string              `json:"type"`
 	Parameters *RuleFileParameters `json:"parameters,omitempty"`
 }
 
 // RepositoryRulesetMaxFilePathLengthRule defines a maximum file path length rule for the repository.
 type RepositoryRulesetMaxFilePathLengthRule struct {
-	//Value for Type: "max_file_path_length"
+	//Type can be one of: "max_file_path_length".
```
Comment (gmlewis) [inline github/repos_rules.go:345]:
```suggestion
	// Type can be one of: "max_file_path_length".
```

File: github/repos_rules.go [hunk 15]
```diff
@@ -335,43 +335,43 @@ type RepositoryRulesetPatternRule struct {
...
 
 	Type       string                           `json:"type"`
 	Parameters *RuleMaxFilePathLengthParameters `json:"parameters,omitempty"`
 }
 
 // RepositoryRulesetFileExtensionRestrictionRule defines a file extension restriction rule for the repository.
 type RepositoryRulesetFileExtensionRestrictionRule struct {
-	//Value for Type: "file_extension_restriction"
+	//Type can be one of: "file_extension_restriction".
```
Comment (gmlewis) [inline github/repos_rules.go:353]:
```suggestion
	// Type can be one of: "file_extension_restriction".
```

File: github/repos_rules.go [hunk 16]
```diff
@@ -335,43 +335,43 @@ type RepositoryRulesetPatternRule struct {
...
 	Type       string                                  `json:"type"`
 	Parameters *RuleFileExtensionRestrictionParameters `json:"parameters,omitempty"`
 }
 
 // RepositoryRulesetMaxFileSizeRule defines a maximum file size rule for the repository.
 type RepositoryRulesetMaxFileSizeRule struct {
-	//Value for Type: "max_file_size"
+	//Type can be one of: "max_file_size".
```
Comment (gmlewis) [inline github/repos_rules.go:360]:
```suggestion
	// Type can be one of: "max_file_size".
```

File: github/repos_rules.go [hunk 17]
```diff
@@ -335,43 +335,43 @@ type RepositoryRulesetPatternRule struct {
...
 	Type       string                     `json:"type"`
 	Parameters *RuleMaxFileSizeParameters `json:"parameters,omitempty"`
 }
 
 // RepositoryRulesetWorkflowsRule defines a workflow rule for the repository.
 type RepositoryRulesetWorkflowsRule struct {
-	//Value for Type: "workflows"
+	//Type can be one of: "workflows".
```
Comment (gmlewis) [inline github/repos_rules.go:367]:
```suggestion
	// Type can be one of: "workflows".
```

File: github/repos_rules.go [hunk 18]
```diff
@@ -335,43 +335,43 @@ type RepositoryRulesetPatternRule struct {
...
 	Type       string                           `json:"type"`
 	Parameters *RequiredWorkflowsRuleParameters `json:"parameters,omitempty"`
 }
 
 // RepositoryRulesetCodeScanningRule defines a code scanning rule for the repository.
 type RepositoryRulesetCodeScanningRule struct {
-	//Value for Type:"code_scanning"
+	//Type can be one of:"code_scanning".
```
Comment (gmlewis) [inline github/repos_rules.go:374]:
```suggestion
	// Type can be one of:"code_scanning".
```

**Review by gmlewis — APPROVED** [review]
Thank you, @unamdev0 and @alexandear !
I'm going to go ahead and commit my tiny nit changes, then we should be good to merge after a second LGTM+Approval from any other contributor to this repo.

File: github/repos_rules.go [hunk 19]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+	Configuration *RepositoryRulesetEditedSources `json:"configuration,omitempty"`
+	RuleType      *RepositoryRulesetEditedSources `json:"rule_type,omitempty"`
+	Pattern       *RepositoryRulesetEditedSources `json:"pattern,omitempty"`
+}
+
+// RepositoryRuleset represents the structure of a ruleset associated with a GitHub repository.
+type RepositoryRuleset struct {
+	ID   int64  `json:"id"`
+	Name string `json:"name"`
+	//Possible values for target: "branch", "tag", "push"
```
Comment (alexandear) [inline github/repos_rules.go:239]:
```suggestion
	// Possible values for target: "branch", "tag", "push"
```

File: github/repos_rules.go [hunk 20]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+	Target *string `json:"target,omitempty"`
+	//Possible values for source type: "Repository", "Organization"
```
Comment (alexandear) [inline github/repos_rules.go:241]:
```suggestion
	// Possible values for source type: "Repository", "Organization"
```

File: github/repos_rules.go [hunk 21]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+	CurrentUserCanBypass *string                  `json:"current_user_can_bypass,omitempty"`
+	NodeID               *string                  `json:"node_id,omitempty"`
+	Links                *RepositoryRulesetLink   `json:"_links,omitempty"`
+	Conditions           json.RawMessage          `json:"conditions,omitempty"`
+	Rules                []*RepositoryRulesetRule `json:"rules,omitempty"`
+	CreatedAt            *Timestamp               `json:"created_at,omitempty"`
+	UpdatedAt            *Timestamp               `json:"updated_at,omitempty"`
+}
+
+// RepositoryRulesetRule represents indivisual rules which are present in a repository's ruleset.
```
Comment (alexandear) [inline github/repos_rules.go:257]:
```suggestion
// RepositoryRulesetRule represents individual rules which are present in a repository's ruleset.
```

File: github/repos_rules.go [hunk 22]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+type RepositoryRulesetRule struct {
+	Creation                 *RepositoryRulesetRuleType                     `json:"creation,omitempty"`
+	Update                   *RepositoryRulesetUpdateRule                   `json:"update,omitempty"`
+	Deletion                 *RepositoryRulesetRuleType                     `json:"deletion,omitempty"`
+	RequireLinearHistory     *RepositoryRulesetRuleType                     `json:"required_linear_history,omitempty"`
```
Comment (alexandear) [inline github/repos_rules.go:262]:
```suggestion
	RequiredLinearHistory    *RepositoryRulesetRuleType                     `json:"required_linear_history,omitempty"`
```

File: github/repos_rules.go [hunk 23]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+	MergeQueue               *RepositoryRulesetMergeQueueRule               `json:"merge_queue,omitempty"`
+	RequireDeployments       *RepositoryRulesetRequireDeploymentsRule       `json:"required_deployments,omitempty"`
```
Comment (alexandear) [inline github/repos_rules.go:264]:
```suggestion
	RequiredDeployments      *RepositoryRulesetRequireDeploymentsRule       `json:"required_deployments,omitempty"`
```

File: github/repos_rules.go [hunk 24]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+// RepositoryRulesetWorkflowsRule defines a workflow rule for the repository.
+type RepositoryRulesetWorkflowsRule struct {
+	// Type can be one of: "workflows".
+	Type       string                           `json:"type"`
+	Parameters *RequiredWorkflowsRuleParameters `json:"parameters,omitempty"`
+}
+
+// RepositoryRulesetCodeScanningRule defines a code scanning rule for the repository.
+type RepositoryRulesetCodeScanningRule struct {
+	// Type can be one of:"code_scanning".
```
Comment (alexandear) [inline github/repos_rules.go:374]:
```suggestion
	// Type can be one of: "code_scanning".
```

**Review by alexandear — CHANGES_REQUESTED** [review]

**Review by gmlewis — COMMENTED** [review]
Ah, thank you for the catches, @alexandear !
I'll apply them.

File: github/event_types_test.go [hunk 25]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
+						Type: "update",
+						Parameters: &UpdateAllowsFetchAndMergeRuleParameters{
+							UpdateAllowsFetchAndMerge: true,
+						},
+					},
+					Deletion: &RepositoryRulesetRuleType{
+						Type: "deletion",
+					},
+					RequireLinearHistory: &RepositoryRulesetRuleType{
+						Type: "require_linear_history",
```
Comment (gmlewis) [inline github/event_types_test.go:9692]:
```suggestion
					RequiredLinearHistory: &RepositoryRulesetRuleType{
						Type: "required_linear_history",
```

File: github/event_types_test.go [hunk 26]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
+							Type: "update",
+							Parameters: &UpdateAllowsFetchAndMergeRuleParameters{
+								UpdateAllowsFetchAndMerge: true,
+							},
+						},
+						Deletion: &RepositoryRulesetRuleType{
+							Type: "deletion",
+						},
+						RequireLinearHistory: &RepositoryRulesetRuleType{
+							Type: "require_linear_history",
```
Comment (gmlewis) [inline github/event_types_test.go:9899]:
```suggestion
						RequiredLinearHistory: &RepositoryRulesetRuleType{
							Type: "required_linear_history",
```

File: github/event_types_test.go [hunk 27]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
```
Comment (gmlewis) [inline github/event_types_test.go:10053]:
```suggestion
						RequiredLinearHistory: &RepositoryRulesetRuleType{
							Type: "required_linear_history",
```

File: github/event_types_test.go [hunk 28]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
+							CheckResponseTimeoutMinutes:  35,
+							GroupingStrategy:             "HEADGREEN",
+							MaxEntriesToBuild:            8,
+							MaxEntriesToMerge:            4,
+							MergeMethod:                  "SQUASH",
+							MinEntriesToMerge:            2,
+							MinEntriesToMergeWaitMinutes: 13,
+						},
+					},
+					RequireDeployments: &RepositoryRulesetRequireDeploymentsRule{
```
Comment (gmlewis) [inline github/event_types_test.go:9706]:
```suggestion
					RequiredDeployments: &RepositoryRulesetRequiredDeploymentsRule{
```

File: github/event_types_test.go [hunk 29]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
+								CheckResponseTimeoutMinutes:  35,
+								GroupingStrategy:             "HEADGREEN",
+								MaxEntriesToBuild:            8,
+								MaxEntriesToMerge:            4,
+								MergeMethod:                  "SQUASH",
+								MinEntriesToMerge:            2,
+								MinEntriesToMergeWaitMinutes: 13,
+							},
+						},
+						RequireDeployments: &RepositoryRulesetRequireDeploymentsRule{
```
Comment (gmlewis) [inline github/event_types_test.go:9913]:
```suggestion
						RequiredDeployments: &RepositoryRulesetRequiredDeploymentsRule{
```

File: github/event_types_test.go [hunk 30]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
```
Comment (gmlewis) [inline github/event_types_test.go:10067]:
```suggestion
						RequiredDeployments: &RepositoryRulesetRequiredDeploymentsRule{
```

File: github/event_types_test.go [hunk 31]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
+									CheckResponseTimeoutMinutes:  35,
+									GroupingStrategy:             "HEADGREEN",
+									MaxEntriesToBuild:            8,
+									MaxEntriesToMerge:            4,
+									MergeMethod:                  "SQUASH",
+									MinEntriesToMerge:            2,
+									MinEntriesToMergeWaitMinutes: 13,
+								},
+							},
+							RequireDeployments: &RepositoryRulesetRequireDeploymentsRule{
```
Comment (gmlewis) [inline github/event_types_test.go:10221]:
```suggestion
							RequiredDeployments: &RepositoryRulesetRequiredDeploymentsRule{
```

File: github/repos_rules.go [hunk 32]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+}
+
+// RepositoryRulesetRule represents individual rules which are present in a repository's ruleset.
+type RepositoryRulesetRule struct {
+	Creation                 *RepositoryRulesetRuleType                     `json:"creation,omitempty"`
+	Update                   *RepositoryRulesetUpdateRule                   `json:"update,omitempty"`
+	Deletion                 *RepositoryRulesetRuleType                     `json:"deletion,omitempty"`
+	RequireLinearHistory     *RepositoryRulesetRuleType                     `json:"required_linear_history,omitempty"`
+	MergeQueue               *RepositoryRulesetMergeQueueRule               `json:"merge_queue,omitempty"`
+	RequireDeployments       *RepositoryRulesetRequireDeploymentsRule       `json:"required_deployments,omitempty"`
```
Comment (gmlewis) [inline github/repos_rules.go:264]:
```suggestion
	RequiredDeployments       *RepositoryRulesetRequiredDeploymentsRule       `json:"required_deployments,omitempty"`
```

File: github/repos_rules.go [hunk 33]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+
+// RepositoryRulesetMergeQueueRule defines a merge queue rule for the repository.
+type RepositoryRulesetMergeQueueRule struct {
+	// Type can be one of: "merge_queue".
+	Type       string                    `json:"type"`
+	Parameters *MergeQueueRuleParameters `json:"parameters,omitempty"`
+}
+
+// RepositoryRulesetRequireDeploymentsRule defines a rule for required deployments.
+type RepositoryRulesetRequireDeploymentsRule struct {
```
Comment (gmlewis) [inline github/repos_rules.go:308]:
```suggestion
// RepositoryRulesetRequiredDeploymentsRule defines a rule for required deployments.
type RepositoryRulesetRequiredDeploymentsRule struct {
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
I'll go ahead and apply these as well.

**Comment by gmlewis** [comment]
Bummer. I didn't get the spacing right. working on it...

File: github/repos_rules.go [hunk 34]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+}
+
+// RepositoryRulesetRule represents individual rules which are present in a repository's ruleset.
+type RepositoryRulesetRule struct {
+	Creation                 *RepositoryRulesetRuleType                     `json:"creation,omitempty"`
+	Update                   *RepositoryRulesetUpdateRule                   `json:"update,omitempty"`
+	Deletion                 *RepositoryRulesetRuleType                     `json:"deletion,omitempty"`
+	RequiredLinearHistory    *RepositoryRulesetRuleType                     `json:"required_linear_history,omitempty"`
+	MergeQueue               *RepositoryRulesetMergeQueueRule               `json:"merge_queue,omitempty"`
+	RequiredDeployments       *RepositoryRulesetRequiredDeploymentsRule       `json:"required_deployments,omitempty"`
```
Comment (gmlewis) [inline github/repos_rules.go:264]:
```suggestion
	RequiredDeployments      *RepositoryRulesetRequiredDeploymentsRule       `json:"required_deployments,omitempty"`
```

File: github/repos_rules.go [hunk 35]
```diff
@@ -167,6 +168,226 @@ type RepositoryRule struct {
...
+}
+
+// RepositoryRulesetRule represents individual rules which are present in a repository's ruleset.
+type RepositoryRulesetRule struct {
+	Creation                 *RepositoryRulesetRuleType                     `json:"creation,omitempty"`
+	Update                   *RepositoryRulesetUpdateRule                   `json:"update,omitempty"`
+	Deletion                 *RepositoryRulesetRuleType                     `json:"deletion,omitempty"`
+	RequiredLinearHistory    *RepositoryRulesetRuleType                     `json:"required_linear_history,omitempty"`
+	MergeQueue               *RepositoryRulesetMergeQueueRule               `json:"merge_queue,omitempty"`
+	RequiredDeployments      *RepositoryRulesetRequiredDeploymentsRule       `json:"required_deployments,omitempty"`
```
Comment (gmlewis) [inline github/repos_rules.go:264]:
```suggestion
	RequiredDeployments      *RepositoryRulesetRequiredDeploymentsRule      `json:"required_deployments,omitempty"`
```

**Comment by gmlewis** [comment]
@unamdev0 - sorry for all the noise - I was trying to make it so that we wouldn't require more changes from you, but alas I was unsuccessful.

Could you please do this for me again?

```
git checkout feature/add-repositoryrulesetevent
git pull origin feature/add-repositoryrulesetevent
go generate ./...
git commit -am 'Run latest go generate'
git push origin feature/add-repositoryrulesetevent
```

**Comment by unamdev0** [comment]
done @gmlewis

File: github/event_types_test.go [hunk 36]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
+                  "type": "update",
+                  "parameters": {
+                    "update_allows_fetch_and_merge": true
+                  }
+                },
+                "deletion": {
+                  "type": "deletion"
+                },
+                "required_linear_history": {
+                  "type": "require_linear_history"
```
Comment (gmlewis) [inline github/event_types_test.go:10550]:
```suggestion
                  "type": "required_linear_history"
```

File: github/event_types_test.go [hunk 37]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
+                    "type": "update",
+                    "parameters": {
+                      "update_allows_fetch_and_merge": true
+                    }
+                  },
+                  "deletion": {
+                    "type": "deletion"
+                  },
+                  "required_linear_history": {
+                    "type": "require_linear_history"
```
Comment (gmlewis) [inline github/event_types_test.go:10783]:
```suggestion
                    "type": "required_linear_history"
```

File: github/event_types_test.go [hunk 38]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
```
Comment (gmlewis) [inline github/event_types_test.go:10944]:
```suggestion
                    "type": "required_linear_history"
```

File: github/event_types_test.go [hunk 39]
```diff
@@ -9506,6 +9506,1783 @@ func TestReleaseEvent_Marshal(t *testing.T) {
...
+                      "type": "update",
+                      "parameters": {
+                        "update_allows_fetch_and_merge": true
+                      }
+                    },
+                    "deletion": {
+                      "type": "deletion"
+                    },
+                    "required_linear_history": {
+                      "type": "require_linear_history"
```
Comment (gmlewis) [inline github/event_types_test.go:11106]:
```suggestion
                      "type": "required_linear_history"
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Unit tests still failing. I'll see if I can fix them.

**Review by gmlewis — APPROVED** [review]
Merging.

**Pull request merged**